File: ml-benchmark/autolifter/single-pass/count1s2s3s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 1;
Config SampleIntMax = 3;

Inductive List = cons {Int, List} | nil Unit;

inf = 100;
two = 2;
three = 3;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

count1s2s3s = fix (
  \f: Bool -> Bool -> List -> Int. \s1: Bool. \s2: Bool. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let upd = if and (== h 3) (or s1 s2) then 1 else 0 in
    let s2 = and (== h 2) (or s1 s2) in
    let s1 = == h 1 in
    + upd (f s1 s2 t)
  end
) false false;

main = single_pass count1s2s3s;

File: ml-benchmark/autolifter/single-pass/count1s2s3s.ml
type l = Nil | Cons of int * l

let inf = 100
let two = 2
let three = 3

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons(h, t) -> Cons(h, run t)
  in
  fun xs -> v (run xs)

let rec count1s2s3s_f s1 s2 xs =
  match xs with
  | Nil -> 0
  | Cons(h, t) ->
    let upd = if h = 3 && (s1 || s2) then 1 else 0 in
    let s2 = (h = 2) && (s1 || s2) in
    let s1 = h = 1 in
    upd + count1s2s3s_f s1 s2 t

let count1s2s3s = count1s2s3s_f false false

let program = single_pass count1s2s3s